src/main/daml/Daml/Finance/Settlement/Instruction.daml

```diff
@@ -69,80 +67,118 @@ template Instruction
+            pure None
           a -> pure None
-        -- verify and disclose allocation
+        -- allocate
         allocationDisclosed <- case allocation of
           Pledge transferableCid -> do
             transferableCid <- discloseT step.sender settlers transferableCid
             transferable <- fetch transferableCid
             let senderAccount = getAccount transferable
+            discloseA senderAccount
             vSenderAccount <- view <$> fetchInterfaceByKey @Account.R senderAccount
-            assertMsg "actors must include transfer instructors" $ S.isSubsetOf vSenderAccount.controllers.instructors actors
+            assertMsg "actors must include instructors" $ S.isSubsetOf vSenderAccount.controllers.instructors actors
             assertMsg "actors must include lockers" $ S.isSubsetOf (getLockers transferable) actors
-            discloseA senderAccount
             getAmount transferable === step.quantity.amount
             getInstrument transferable === step.quantity.unit
             pure $ Pledge transferableCid
+          PassthroughFrom (passthroughAccount, fromInstruction) -> do
+            discloseA passthroughAccount
+            vSenderAccount <- view <$> fetchInterfaceByKey @Account.R passthroughAccount
+            assertMsg "actors must include instructors" $ S.isSubsetOf vSenderAccount.controllers.instructors actors
+            pure $ PassthroughFrom (passthroughAccount, fromInstruction)
           a -> do
             assertMsg "actors must include sender" $ S.member step.sender actors
             pure a
-        -- update instruction state
         cid <- toInterfaceContractId <$> create this with allocation = allocationDisclosed; signed = S.insert step.sender signed
         pure (cid, otCid)
-      approve Instruction.Approve{approval; actors} = do
+      approve Instruction.Approve{actors; approval} = do
+        let newSigned = S.insert step.receiver signed
         assertMsg ("approval must be new" <> messageSuffix) $ approval /= this.approval
-        -- clean up if previously approved
+        -- undisclose previous approval
         case this.approval of
           TakeDelivery receiverAccount -> undiscloseA receiverAccount
+          PassthroughTo (passthroughAccount, _) -> undiscloseA passthroughAccount
           a -> pure None
-        -- disclose approval
+        -- approve
         case approval of
           TakeDelivery receiverAccount -> do
-            vReceiverAccount <- view <$> fetchInterfaceByKey @Account.R receiverAccount
             discloseA receiverAccount
-            assertMsg "actors must include transfer approvers" $ S.isSubsetOf vReceiverAccount.controllers.approvers actors
+            vReceiverAccount <- view <$> fetchInterfaceByKey @Account.R receiverAccount
+            assertMsg "actors must include approvers" $ S.isSubsetOf vReceiverAccount.controllers.approvers actors
+          PassthroughTo (passthroughAccount, toInstruction) -> do
```

Comment:
Not for this PR, but I think that once the `receiver` approves with a `PassThroughTo`, we could directly exercise `Allocate` on the `toInstruction` with a `PassThroughFrom`?

Reply:
I initially did that, but decided to rip it out. It added quite some complexity to make sure the instructions stay in sync (especially in case of reapprovals/reallocations). I was thinking of making sure that a `PassthroughFrom` allocation only happens through the `PassthroughTo` approval code (with a call token or so) which might have reduced the complexity somewhat. Happy to reconsider it in a separate PR.